File: implementations/java/client/src/main/java/com/theshuai/tunnelclient/peer/PeerAppMessageCodec.java
package com.theshuai.tunnelclient.peer;

import com.theshuai.common.util.JsonUtil;
import lombok.Data;
import org.springframework.util.StringUtils;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;

final class PeerAppMessageCodec {
    static final String TYPE_MESSAGE = "message";
    static final String TYPE_ACK = "ack";
    private static final byte[] PREFIX = "STMSG2\n".getBytes(StandardCharsets.US_ASCII);

    private PeerAppMessageCodec() {
    }

    static boolean looksLike(byte[] payload) {
        return payload != null
                && payload.length >= PREFIX.length
                && Arrays.equals(PREFIX, Arrays.copyOfRange(payload, 0, PREFIX.length));
    }

    static PeerAppMessage decode(byte[] payload) {
        if (!looksLike(payload)) {
            return null;
        }
        String json = new String(payload, PREFIX.length, payload.length - PREFIX.length, StandardCharsets.UTF_8);
        PeerAppMessage message = JsonUtil.stringToObject(json, PeerAppMessage.class);
        return message != null && StringUtils.hasText(message.getType()) ? message : null;
    }

    static byte[] encode(PeerAppMessage message) {
        String json = JsonUtil.objectToString(message);
        byte[] body = (json == null ? "{}" : json).getBytes(StandardCharsets.UTF_8);
        byte[] payload = new byte[PREFIX.length + body.length];
        System.arraycopy(PREFIX, 0, payload, 0, PREFIX.length);
        System.arraycopy(body, 0, payload, PREFIX.length, body.length);
        return payload;
    }

    @Data
    static final class PeerAppMessage {
        private String type;
        private String id;
        private long fromClientId;
        private String fromClientName;
        private long toClientId;
        private String toClientName;
        private String message;
        private long createdAtMillis;
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerAppMessageCodec.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Specus.Client.PeerMesh;

internal static class PeerAppMessageCodec
{
    public const string TypeMessage = "message";
    public const string TypeAck = "ack";

    private static readonly byte[] Prefix = Encoding.ASCII.GetBytes("STMSG2\n");

    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault,
    };

    public static bool LooksLike(byte[] payload)
    {
        return payload.Length >= Prefix.Length
            && payload.AsSpan(0, Prefix.Length).SequenceEqual(Prefix);
    }

    public static bool TryDecode(byte[] payload, out PeerAppMessage message)
    {
        message = default!;
        if (!LooksLike(payload))
        {
            return false;
        }
        try
        {
            var decoded = JsonSerializer.Deserialize<PeerAppMessage>(
                payload.AsSpan(Prefix.Length),
                JsonOptions);
            if (decoded is null || string.IsNullOrWhiteSpace(decoded.Type))
            {
                return false;
            }
            message = decoded;
            return true;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    public static byte[] Encode(PeerAppMessage message)
    {
        var wireMessage = string.Equals(message.Type, TypeMessage, StringComparison.OrdinalIgnoreCase)
            ? message : WithoutAttachment(message);
        var json = JsonSerializer.SerializeToUtf8Bytes(wireMessage, JsonOptions);
        var payload = new byte[Prefix.Length + json.Length];
        Prefix.CopyTo(payload, 0);
        json.CopyTo(payload.AsSpan(Prefix.Length));
        return payload;
    }

    public static string DisplayText(PeerAppMessage message)
    {
        if (message.Attachment is null)
        {
            return message.Message ?? "";
        }
        var size = message.Attachment.SizeBytes > 0 ? FormatBytes(message.Attachment.SizeBytes) : "-";
        var text = string.IsNullOrWhiteSpace(message.Message) ? "" : $"{message.Message} ";
        return $"{text}[附件] {FirstNonEmpty(message.Attachment.FileName, message.Attachment.ObjectId, "attachment")} · {FirstNonEmpty(message.Attachment.MimeType, "application/octet-stream")} · {size}";
    }

    private static PeerAppMessage WithoutAttachment(PeerAppMessage message) => new()
    {
        Type = message.Type,
        Id = message.Id,
        FromClientId = message.FromClientId,
        FromClientName = message.FromClientName,
        ToClientId = message.ToClientId,
        ToClientName = message.ToClientName,
        Message = message.Message,
        CreatedAtMillis = message.CreatedAtMillis,
    };

    private static string FormatBytes(long bytes)
    {
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        var value = (double)Math.Max(0L, bytes);
        var unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }
        return unit == 0 ? $"{value:0} {units[unit]}" : $"{value:0.##} {units[unit]}";
    }

    private static string FirstNonEmpty(params string?[] values)
    {
        return values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value))?.Trim() ?? "";
    }
}

internal sealed class PeerAppMessage
{
    [JsonPropertyName("type")]
    public string? Type { get; set; }

    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("fromClientId")]
    public long FromClientId { get; set; }

    [JsonPropertyName("fromClientName")]
    public string? FromClientName { get; set; }

    [JsonPropertyName("toClientId")]
    public long ToClientId { get; set; }

    [JsonPropertyName("toClientName")]
    public string? ToClientName { get; set; }

    [JsonPropertyName("message")]
    public string? Message { get; set; }

    [JsonPropertyName("attachment")]
    public PeerAppAttachment? Attachment { get; set; }

    [JsonPropertyName("createdAtMillis")]
    public long CreatedAtMillis { get; set; }
}

internal sealed class PeerAppAttachment
{
    [JsonPropertyName("objectId")]
    public string? ObjectId { get; set; }

    [JsonPropertyName("attachmentId")]
    public long AttachmentId { get; set; }

    [JsonPropertyName("fileName")]
    public string? FileName { get; set; }

    [JsonPropertyName("mimeType")]
    public string? MimeType { get; set; }

    [JsonPropertyName("sizeBytes")]
    public long SizeBytes { get; set; }

    [JsonPropertyName("sha256")]
    public string? Sha256 { get; set; }
}
